embassy-time/src/queue_generic.rs

```diff
@@ -0,0 +1,453 @@
+use core::cell::RefCell;
+use core::cmp::Ordering;
+use core::task::Waker;
+
+use embassy_sync::blocking_mutex::raw::CriticalSectionRawMutex;
+use embassy_sync::blocking_mutex::Mutex;
+use heapless::sorted_linked_list::{LinkedIndexU8, Min, SortedLinkedList};
+
+use crate::driver::{allocate_alarm, set_alarm, set_alarm_callback, AlarmHandle};
+use crate::queue::TimerQueue;
+use crate::Instant;
+
+#[cfg(feature = "generic-queue-8")]
+const QUEUE_SIZE: usize = 8;
+#[cfg(feature = "generic-queue-16")]
+const QUEUE_SIZE: usize = 16;
+#[cfg(feature = "generic-queue-32")]
+const QUEUE_SIZE: usize = 32;
+#[cfg(feature = "generic-queue-64")]
+const QUEUE_SIZE: usize = 32;
+#[cfg(feature = "generic-queue-128")]
+const QUEUE_SIZE: usize = 128;
+#[cfg(not(any(
+    feature = "generic-queue-8",
+    feature = "generic-queue-16",
+    feature = "generic-queue-32",
+    feature = "generic-queue-64",
+    feature = "generic-queue-128"
+)))]
+const QUEUE_SIZE: usize = 64;
+
+#[derive(Debug)]
+struct Timer {
+    at: Instant,
+    waker: Waker,
+}
+
+impl PartialEq for Timer {
+    fn eq(&self, other: &Self) -> bool {
+        self.at == other.at
+    }
+}
+
+impl Eq for Timer {}
+
+impl PartialOrd for Timer {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.at.partial_cmp(&other.at)
+    }
+}
+
+impl Ord for Timer {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.at.cmp(&other.at)
+    }
+}
+
+struct InnerQueue {
+    queue: SortedLinkedList<Timer, LinkedIndexU8, Min, { QUEUE_SIZE }>,
+    alarm: Option<AlarmHandle>,
+    alarm_at: Instant,
+}
+
+impl InnerQueue {
+    const fn new() -> Self {
+        Self {
+            queue: SortedLinkedList::new_u8(),
+            alarm: None,
+            alarm_at: Instant::MAX,
+        }
+    }
+
+    fn schedule_wake(&mut self, at: Instant, waker: &Waker) {
+        self.queue
+            .find_mut(|timer| timer.waker.will_wake(waker))
+            .map(|mut timer| {
+                timer.at = at;
```

Comment:
this needs to keep the min of old and new `at`s, not overwrite

Reply:
If you can elaborate why, that would be great! I can't see why myself.

Reply:
A single task is running two different futures with `join`, each waits on a different timer. Fut1 waits for instant 50, fut2 waits for instant 100. When polling the task, `join` polls fut1 then fut2. 

Polling fut2 will overwrite `at=100`, so the task will be polled at instant `100`, but we want it to get polled at 50.

Reply:
Agreed! Will fix. 

(I thought for a second we have a bigger problem if a timer future is dropped and a new timer future is created by the same task, but fortunately, each time a timer is polled, it reschedules itself so your "min" would work just fine.)